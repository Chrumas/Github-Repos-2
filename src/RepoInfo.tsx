import React from "react";
import { useState, useEffect } from "react";
import { openInNewTab } from "./UserInfo";
import { StyledRepoInfo } from "./styles/RepoInfo.styled";
import { ErrorMessage } from "./ErrorMessage";

interface props {
  url: string;
  name: string;
  stargazers: string;
  owner: string;
}

export const RepoInfo = ({ url, name, stargazers, owner }: props) => {
  const [languages, setLanguages] = useState<any>();
  const [showLanguages, setShowLanguages] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState("");

  //Fetches additional data from the server (about langauges used)
  useEffect(() => {
    var tempArray: string[] = [];
    let languagesUrl =
      "https://api.github.com/repos/" + owner + "/" + name + "/languages";
    try {
      fetch(languagesUrl)
        .then((response) => {
          if (!response.ok) {
            if (response.status === 403) {
              setErrorMessage("API limit exceeded!");
              throw new Error("API limit exceeded!");
            }
          }
          return response.json();
        })
        .then((data) => {
          setErrorMessage("");
          for (let i in Object.keys(data)) {
            if (Object.keys(data)[i] !== "undefined")
              tempArray.push(
                Object.keys(data)[i] + ": " + Object.values(data)[i] + " bytes"
              );
          }
          var map = tempArray.map((choice: any, index: number) => {
            return (
              <>
                <div>{choice}</div>
              </>
            );
          });
          setLanguages(map);
        })
        .catch((e) => {
          console.log(e);
        });
    } catch {}
  }, [name, owner]);

  //Slices the printable url and name if it's too long
  var tempUrl = url;
  var tempName = name;
  if (url.length >= 40) {
    tempUrl = tempUrl.slice(0, 39) + "...";
  }
  if (name.length >= 30) {
    tempName = tempName.slice(0, 29) + "...";
  }

  if (showLanguages === true) {
    return (
      <StyledRepoInfo>
        <div className="repository">
          <div>{tempName}</div>
          <div onClick={() => openInNewTab(url)} className="link">
            Link: {tempUrl}
          </div>
          <div>{"Amount of stars: " + stargazers}</div>
          <br></br>
          <div>{languages}</div>
          <div
            onClick={() => setShowLanguages(false)}
            className="clickable-text"
          >
            Show less
          </div>
        </div>
      </StyledRepoInfo>
    );
  } else {
    return (
      <StyledRepoInfo>
        <div className="repository">
          <div>{tempName}</div>
          <div onClick={() => openInNewTab(url)} className="link">
            Link: {tempUrl}
          </div>
          <div>{"Amount of stars: " + stargazers}</div>
          <br></br>
          <div
            onClick={() => setShowLanguages(true)}
            className="clickable-text"
          >
            More
          </div>
          <ErrorMessage inputName={errorMessage} inputColor="red" />
        </div>
      </StyledRepoInfo>
    );
  }
};
