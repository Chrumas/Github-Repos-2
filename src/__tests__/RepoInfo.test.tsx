import * as React from "react";
import * as ReactDOM from "react-dom";
import { RepoInfo } from "../RepoInfo";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <RepoInfo url="any" name="any" stargazers="non" owner="uwu" />,
    div
  );
});

it("matches snapshot", () => {
  const tree = renderer
    .create(<RepoInfo url="any" name="any" stargazers="non" owner="uwu" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
