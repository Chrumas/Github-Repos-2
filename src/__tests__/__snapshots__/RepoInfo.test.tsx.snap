// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-gsDKAQ flOrLh"
>
  <div
    className="repository"
  >
    <div>
      any
    </div>
    <div
      className="link"
      onClick={[Function]}
    >
      Link: 
      any
    </div>
    <div>
      Amount of stars: non
    </div>
    <br />
    <div
      className="clickable-text"
      onClick={[Function]}
    >
      More
    </div>
  </div>
</div>
`;
